### ISSUE TITLE
LUKS2 (LVM on LUKS) volume not unlocked with TPM key

### ISSUE THREAD

**Issue by bitthief** [dracutdevs/dracut#1729]
**Describe the bug**
Dracut is not unlocking the LUKS2 (LVM on LUKS) volume with the pre-enrolled TPM key (in slot 2) and asks for an unlock passphrase (in slot 1) instead.

Additionally, TPM2 TOTP errors are printed:

![image](https://user-images.githubusercontent.com/29442776/156025716-ecb9f39f-986f-473d-b4cf-e149dec7660a.png)

**Distribution used**
Gentoo Linux (VM, Hyper-V Generation 2/EFI VM with vTPM), systemd-boot bootloader installed in ESP

**Dracut version**
056-8-g4c355d05

**Init system**
systemd

**To Reproduce**
`systemd-cryptenroll --tpm2-device=auto --tpm2-pcrs=0+1+2+3+4+5+7+8 /dev/sda2`

**Expected behavior**
LUKS2 volume should be unlocked without user interaction at boot.

**Additional context**

**Kernel cmdline:**
Linux version 5.16.11-gentoo (root@0xffff) (gcc (Gentoo Hardened 11.2.1_p20220115 p4) 11.2.1 20220115, GNU ld (Gentoo 2.37_p1 p2) 2.37) #1 SMP PREEMPT Sun Feb 27 08:37:04 EET 2022
Command line: **root=UUID=664420b6-189b-48f4-b42c-2293ed000fb9 rd.luks.uuid=edab8e80-39ce-4b57-bade-3eae65c02034 rd.luks.options=edab8e80-39ce-4b57-bade-3eae65c02034=tpm2-device=auto,discard** rd.lvm.vg=vg1 rd.luks.allow-discards net.ifnames=0 transparent_hugepage=madvise scsi_mod.use_blk_mq=1 dm_mod.use_blk_mq=1 video=hyperv_fb:1920x1080

**cryptsetup luksDump /dev/sda2**
```
LUKS header information
Version:       	2
Epoch:         	6
Metadata area: 	16384 [bytes]
Keyslots area: 	16744448 [bytes]
UUID:          	edab8e80-39ce-4b57-bade-3eae65c02034
Label:         	(no label)
Subsystem:     	(no subsystem)
Flags:       	(no flags)

Data segments:
  0: crypt
	offset: 16777216 [bytes]
	length: (whole device)
	cipher: serpent-xts-plain64
	sector: 512 [bytes]

Keyslots:
  0: luks2
	Key:        512 bits
	Priority:   normal
	Cipher:     serpent-xts-plain64
	Cipher key: 512 bits
	PBKDF:      argon2i
	Time cost:  10
	Memory:     1048576
	Threads:    4
	Salt:       f2 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            b1 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	AF stripes: 4000
	AF hash:    whirlpool
	Area offset:32768 [bytes]
	Area length:258048 [bytes]
	Digest ID:  0
  1: luks2
	Key:        512 bits
	Priority:   normal
	Cipher:     serpent-xts-plain64
	Cipher key: 512 bits
	PBKDF:      argon2i
	Time cost:  10
	Memory:     1048576
	Threads:    4
	Salt:       92 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            af ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	AF stripes: 4000
	AF hash:    sha256
	Area offset:290816 [bytes]
	Area length:258048 [bytes]
	Digest ID:  0
  2: luks2
	Key:        512 bits
	Priority:   normal
	Cipher:     serpent-xts-plain64
	Cipher key: 512 bits
	PBKDF:      pbkdf2
	Hash:       sha512
	Iterations: 1000
	Salt:       1a ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            77 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	AF stripes: 4000
	AF hash:    sha512
	Area offset:548864 [bytes]
	Area length:258048 [bytes]
	Digest ID:  0
Tokens:
  0: systemd-tpm2
	tpm2-pcrs:  0, 1, 2, 3, 4, 5, 7, 8
	tpm2-bank:  sha256
	tpm2-primary-alg:  ecc
	tpm2-blob:  00 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            cc ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            e9 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            94 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            09 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            53 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            6f ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            94 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            00 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            f3 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            b9 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            49 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            2a ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	tpm2-policy-hash:
	            44 ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	            db ** ** ** ** ** ** ** ** ** ** ** ** ** ** **
	Keyslot:    2
```

**cat /etc/dracut.conf.d/gentoo.conf**
```
add_dracutmodules+=" lvm crypt crypt-gpg drm selinux systemd plymouth virtfs tpm2-tss tpm2-totp "

install_items+=" /usr/lib64/cryptsetup/libcryptsetup-token-systemd-tpm2.so /usr/lib64/cryptsetup/libcryptsetup-token-systemd-pkcs11.so "

add_drivers+=" fbdev hv_vmbus hv_sock hv_utils hv_balloon hv_netvsc hv_storvsc hyperv_drm hyperv_keyboard hid_hyperv pci_hyperv uio_hv_generic virtio_pci virtio_mem dxgkrnl virtiofs exfat vfat lz4 lz4_compress zstd zram zswap tpm tpm_crb tpm_tis "

add_device+=" /dev/mapper/vg1-swap "

use_fstab="yes"
show_modules="yes"
compress="xz"

uefi="yes"
early_microcode="yes"
uefi_stub="/usr/lib/systemd/boot/efi/linuxx64.efi.stub"

uefi_secureboot_cert="/etc/**/**.crt"
uefi_secureboot_key="/etc/**/**.key"
```

**lsblk -o NAME,UUID,PARTUUID,TYPE**
```
NAME                                          UUID                                   PARTUUID                             TYPE
sda                                                                                                                       disk
|-sda1                                        50A4-90C2                              eacdd9fb-cab0-43de-9f54-ca6232de9161 part
`-sda2                                        edab8e80-39ce-4b57-bade-3eae65c02034   d431524a-d01c-4fea-9225-f80a2681cb67 part
  `-luks-edab8e80-39ce-4b57-bade-3eae65c02034 WNIeug-JhV1-g8d8-XDOi-qe6u-H97L-YQe6td                                      crypt
    |-vg1-swap                                0b1fb29f-8644-405b-8edc-d60b6159f102                                        lvm
    `-vg1-root                                664420b6-189b-48f4-b42c-2293ed000fb9                                        lvm
```

**Comment by AdelKS** [comment]
I had an issue with my tpm2 not being used to unlock my luks partition (without LVM) and after hours of trying to figure this out I found the culprit : I had to add the needed kernel modules for tpm in `/etc/modules-load.d` to get them early loaded from the initramfs. Hopefully that fixes your issue too.

**Comment by bitthief** [comment]
@AdelKS thanks a lot for your hints, I finally got back to fixing this a few days ago.

While it wasn't the same problem for me, you did point me to the correct answer.

Digging, I found a few Hyper-V vTPM issues, which luckily weren't affecting me:
https://github.com/tpm2-software/tpm2-tss/issues/1550
https://github.com/tpm2-software/tpm2-tools/issues/2302


There were a few different smaller issues though:

```
systemd-cryptenroll --tpm2-device=list
PATH        DEVICE      DRIVER
**/dev/tpmrm0** VTPM0101:00 tpm_crb
```

The /dev/tpmrm0 device was not present/usable in the early boot/dracut initramfs phase. I added it to the **add_devices** line. This is a snippet of a working dracut conf:

```
add_dracutmodules+=" lvm crypt crypt-gpg drm selinux systemd plymouth virtfs tpm2-tss tpm2-totp "

add_drivers+=" fbdev hv_vmbus hv_sock hv_utils hv_balloon hv_netvsc hv_storvsc hyperv_drm hyperv_keyboard hid_hyperv pci_hyperv uio_hv_generic virtio_pci virtio_mem dxgkrnl virtiofs exfat vfat lz4 lz4_compress zstd zram zswap tpm tpm_crb tpm_tis tpm_vtpm_proxy "

add_device+=" /dev/mapper/vg1-swap /dev/tpmrm0 "
```

I also added the explicit device to the kernel boot cmdline, instead of auto:

`rd.luks.options=edab8e80-39ce-4b57-bade-3eae65c02034=tpm2-device=**/dev/tpmrm0**,discard`

And, just to be sure, I also wiped the LUKS TPM slot and re-created it with systemd:

```
sudo systemd-cryptenroll /dev/sda2 --wipe-slot=tpm2
sudo systemd-cryptenroll --tpm2-device=/dev/tpmrm0 --tpm2-pcrs=0+7 /dev/sda2
```

After this, I can get dracut to successfully decrypt the LUKS LVM2 volume with no intervention. There's still a random TPM firmware bug error earlier, but everything is working perfectly now!
